### ISSUE TITLE
Add full text search support to Oxigraph

### ISSUE THREAD

**Issue by Tpt** [oxigraph/oxigraph#48]
It would be nice to have some kind of full text search support in Oxigraph.
On the query level, it should probably be exposed as a special `SERVICE` to avoid having to tweak SPARQL grammar similarly to [blazegraph](https://github.com/blazegraph/database/wiki/FullTextSearch).

On the implementation side, I would be more in inclined to write our own search system using existing components like [the Snowball stemmer](https://snowballstem.org/) in order to keep having a single storage system (Sled/RocksDB). It would make things like transactions much simpler to implement.

**Comment by ktk** [comment]
There is no consensus yet but it's one of the topics discussed in the SPARQL 1.2 CWG: https://github.com/w3c/sparql-12/issues/40

Pretty much all implementations I've seen so far are using Lucene but that would drag the whole JVM stack with it I suppose.

**Comment by Tpt** [comment]
@ktk Yes, pulling a JVM does not seem a good way to go. The "search in Rust" area is quite active with [Tantivy](https://github.com/tantivy-search/tantivy) and [MeiliSearch](https://github.com/meilisearch/MeiliSearch).

**Comment by epoz** [comment]
How about using the FTS support of sqlite? https://www.sqlite.org/fts5.html

**Comment by ktk** [comment]
Just stumbled over this today. Will play with it the next weeks as we need a generic text search solution for facet stuff, SPARQL is not that great for it right now (paging, order, etc...): https://blog.meilisearch.com/v1-enterprise-ready-stable/

According to the Github "Languages" information the search-part is pretty much all Rust, might be something to check out: https://github.com/meilisearch/meilisearch.

Not that this is priority now but just to make sure it's known here :)

**Comment by ktk** [comment]
Eh never mind, I just see you *did* mention this @Tpt , sorry about ;)

**Comment by nikokaoja** [comment]
What is the status of free text  search in oxi?

**Comment by Tpt** [comment]
@nikokaoja Sadly no progress until now. Help welcome!

**Comment by bollwyvl** [comment]
Perhaps as a stopgap today, the new [custom functions](https://github.com/oxigraph/oxigraph/blob/v0.4.6/python/tests/test_store.py#L198) (does this work in JS as well?) could create and then work with an index provided by another tool. 

I guess one might need to define:
- maybe a structure that defined the pattern for building an index
  - simplest would be a dumb config blob literal
  - more interesting might be a more structured description of the index(es) to build 
- some functions that (re-)built the index
- some functions that could be used in where/filter/sort to actually limit queries

**Comment by epoz** [comment]
FWIW, I do FTS in oxigraph by doing query re-writing, before passing the query on to the endpoint.

See https://shmarql.com/fizzysearch/    - this is not oxigraph specific per se, but I use pyoxigraph as the preferred triplestore in shmarql, and it works really well in this way.

In this way you get the best of both worlds, proper SPARQL support from oxigraph, and easy/fast FTS from sqlite.
